<a href="https://colab.research.google.com/github/leelasankar/LLMOps/blob/main/MLOps.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#MLOps

<img class="mimg" style="color: rgb(40, 72, 115);" height="180" width="272" src="https://th.bing.com/th/id/OIP.79E4A_2H2qLPlY8LaxIvzwHaEK?w=272&amp;h=180&amp;c=7&amp;r=0&amp;o=5&amp;dpr=1.1&amp;pid=1.7" alt="Image result for mlops" data-thhnrepbd="1" data-bm="155">

# Step 1 : Install Required packages

In [ ]:
!mlflow

/bin/bash: line 1: mlflow: command not found


In [ ]:
!pip install mlflow

In [ ]:
!pip install scikit-learn  # Install scikit-learn if you haven't already

from sklearn.model_selection import train_test_split  # Import the function



In [ ]:
!mlflow

Usage: mlflow [OPTIONS] COMMAND [ARGS]...

Options:
  --version  Show the version and exit.
  --help     Show this message and exit.

Commands:
  artifacts    Upload, list, and download artifacts from an MLflow...
  db           Commands for managing an MLflow tracking database.
  deployments  Deploy MLflow models to custom targets.
  doctor       Prints out useful information for debugging issues with MLflow.
  experiments  Manage experiments.
  gc           Permanently delete runs in the `deleted` lifecycle stage.
  models       Deploy MLflow models locally.
  recipes      Run MLflow Recipes and inspect recipe results.
  run          Run an MLflow project from the given URI.
  runs         Manage runs.
  sagemaker    Serve models on SageMaker.
  server       Run the MLflow tracking server.


In [ ]:
!mlflow --version

mlflow, version 2.15.1


# Import required Packages

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import Elastic
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import mlflow
import mlflow.sklearn
import predicted_data = lr.predict(test_x) # define predicted_data


$Step-3$: **Read the data**

In [ ]:
new_var = "/content/hour.csv"
import pandas as pd
data=pd.read_csv(new_var)
data.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1   0        0        6           0   
1        2  2011-01-01       1   0     1   1        0        6           0   
2        3  2011-01-01       1   0     1   2        0        6           0   
3        4  2011-01-01       1   0     1   3        0        6           0   
4        5  2011-01-01       1   0     1   4        0        6           0   

   weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
0           1  0.24  0.2879  0.81        0.0       3          13   16  
1           1  0.22  0.2727  0.80        0.0       8          32   40  
2           1  0.22  0.2727  0.80        0.0       5          27   32  
3           1  0.24  0.2879  0.75        0.0       3          10   13  
4           1  0.24  0.2879  0.75        0.0       0           1    1

# It is a supervised learning
# it is regression problem

In [ ]:
data.shape

(17379, 17)

 # This dataset has 17379 rows and 17 columns
 # 16 columns are input columns (x)
 # 1 column is i.e., count is output/target variable (y)

In [ ]:
data.columns

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='object')

# My interest is model development and deployment
# to develop model, make sure we have all the columns are numerical data
# math works only on numbers


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


In [ ]:
data.isnull().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

$Step:$ **Model Development using MLflow

- We devide data in two parts train_data and test_data
- We devide train_data in to X_train and y_train
- We devide test_data into X-test and y_test
- Model development happens on train_data
- Model prediction happens on X_test data, this will give y_prediction
- Finally I will compare y predictions with y_test


In [ ]:
import pandas as pd
data = pd.read_csv('hour.csv')

# Assuming 'data' is now your Pandas DataFrame and 'cnt' is the name of your target variable column
X = data.drop('cnt',axis=1)  # Features (all columns except the target)
y = data['cnt']  # Target variable

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
train, test=train_test_split(data,test_size=0.2,random_state=42)
train.shape,test.shape

((13903, 17), (3476, 17))

In [ ]:
train.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  workingday  \
335       336  2011-01-15       1   0     1  11        0        6           0   
7035     7036  2011-10-25       4   0    10  18        0        2           1   
8051     8052  2011-12-07       4   0    12   3        0        3           1   
2133     2134  2011-04-03       2   0     4  18        0        0           0   
8485     8486  2011-12-25       1   0    12   6        0        0           0   

      weathersit  temp   atemp   hum  windspeed  casual  registered  cnt  
335            1  0.20  0.1970  0.55     0.2239      18          54   72  
7035           1  0.52  0.5000  0.42     0.1045      65         453  518  
8051           3  0.46  0.4545  1.00     0.2239       0           3    3  
2133           2  0.46  0.4545  0.31     0.0000      71         101  172  
8485           1  0.20  0.2273  0.75     0.1045       0           1    1

In [ ]:
#Now divide data into train_ and trainy
train_x=train.drop(['cnt'],axis=1)
test_x=test.drop(['cnt'],axis=1)
train_y=train['cnt']
test_y=test['cnt']

In [ ]:
train_x.shape,train_y.shape

((13903, 16), (13903,))

$step-5$: **Start mlflow**

In [ ]:
# mlflow is cloud platform
#I want to develop a model, deploy on mlfow
# Leela Sankar creat a model -----mlflow
# MLflow will create some workspace in the name of experiment

In [ ]:
!pip install mlflow

In [ ]:
mlflow.set_experiment("/mlflow/mlops_experiment")

<Experiment: artifact_location='file:///content/mlruns/799115071868702862', creation_time=1723411109400, experiment_id='799115071868702862', last_update_time=1723411109400, lifecycle_stage='active', name='/mlflow/mlops_experiment', tags={}>

In [ ]:
# prompt: train_model(0.3,0.02)

train_model(0.3,0.02)


RMSE: 0.006553391456857675
MAE: 0.004225487887490494
R2: 0.9999999986895329


Registered model 'Elastic' already exists. Creating a new version of this model...
Created version '3' of model 'Elastic'.
2024/08/11 22:11:06 WARNING mlflow.models.model: Input example should be provided to infer model signature if the model signature is not provided when logging the model.


In [ ]:
print(mlflow.set_experiment("/mlflow/Bike Sharing Demand").experiment_id)
print(mlflow.set_experiment("/mlflow/Bike Sharing Demand").lifecycle_stage)
print(mlflow.set_experiment("/mlflow/Bike Sharing Demand").name)

866567845166838322
active
/mlflow/Bike Sharing Demand


In [ ]:
def train_model(alpha,l1_ratio):
  #develop train test split
  train, test=train_test_split(data,test_size=0.2,random_state=1234)
  train_x=train.drop(['cnt'],axis=1)
  test_x=test.drop(['cnt'],axis=1)
  train_y=train['cnt']
  test_y=test['cnt']

  # ========================= Handle Date/String Columns =========================
  # Identify columns with 'object' dtype (likely strings or dates)
  object_columns = train_x.select_dtypes(include=['object']).columns

  # Convert these columns to numerical representations (example using one-hot encoding)
  for col in object_columns:
      train_x = pd.concat([train_x, pd.get_dummies(train_x[col], prefix=col)], axis=1)
      train_x.drop(col, axis=1, inplace=True)

      test_x = pd.concat([test_x, pd.get_dummies(test_x[col], prefix=col)], axis=1)
      test_x.drop(col, axis=1, inplace=True)

  # Ensure both train_x and test_x have the same columns after one-hot encoding
  missing_cols_train = set(test_x.columns) - set(train_x.columns)
  for col in missing_cols_train:
      train_x[col] = 0
  missing_cols_test = set(train_x.columns) - set(test_x.columns)
  for col in missing_cols_test:
      test_x[col] = 0
  train_x = train_x[test_x.columns]  # Reorder columns to match

  #=================Initiate mlflow==================
  with mlflow.start_run(experiment_id=866567845166838322,run_name='Regression',description='performing regression analysis'):
    #===========================Call ml model====================
    lr=ElasticNet(alpha=alpha,l1_ratio=l1_ratio)
    lr.fit(train_x,train_y)
    #=============Model Prediction
    predicted_data=lr.predict(test_x)
    # Model Evluation============================
    rmse=np.sqrt(mean_squared_error(test_y,predicted_data))
    mae=mean_absolute_error(test_y,predicted_data)
    r2=r2_score(test_y,predicted_data)
    print("RMSE:",rmse)
    print("MAE:",mae)
    print("R2:",r2)

    #log the metrics, parameters and model mlflow
    mlflow.log_param("alpha",alpha)
    mlflow.log_param("l1_ratio",l1_ratio)

    mlflow.log_metric("rmse",rmse)
    mlflow.log_metric("mae",mae)
    mlflow.log_metric("r2",r2)

    mlflow.sklearn.log_model(lr,"model",registered_model_name="Elastic")

In [ ]:
train_model(0.3,0.02)

RMSE: 0.006553391456857675
MAE: 0.004225487887490494
R2: 0.9999999986895329


Registered model 'Elastic' already exists. Creating a new version of this model...
Created version '2' of model 'Elastic'.
2024/08/11 21:54:48 WARNING mlflow.models.model: Input example should be provided to infer model signature if the model signature is not provided when logging the model.


#Step - 6 MlflowUi
#To create Mlflow ui from google colob notebook
#We need to create a tunnel
#Tunnel name ======ngrok


In [ ]:
!pip install pyngrok

In [ ]:
from pyngrok import ngrok

# Get your authtoken from https://dashboard.ngrok.com/auth
ngrok.set_auth_token("YOUR_AUTHTOKEN_HERE")

$ Create a Tunnel$

In [ ]:
auth_token="2kX3XlC674O10DIULmmuY7M2Jsf_3bvjAyEFaSR2EPQb4hgC9"
ngrok.set_auth_token(auth_token)
ngrok_tunnel = ngrok.connect(addr="5000", proto="http",)
print("Tracking uri:", ngrok_tunnel.public_url)

Tracking uri: https://02f2-107-167-189-250.ngrok-free.app


In [ ]:
!mlflow ui

[2024-08-11 23:17:47 +0000] [78733] [INFO] Starting gunicorn 22.0.0
[2024-08-11 23:17:47 +0000] [78733] [INFO] Listening at: http://127.0.0.1:5000 (78733)
[2024-08-11 23:17:47 +0000] [78733] [INFO] Using worker: sync
[2024-08-11 23:17:47 +0000] [78734] [INFO] Booting worker with pid: 78734
[2024-08-11 23:17:47 +0000] [78735] [INFO] Booting worker with pid: 78735
[2024-08-11 23:17:47 +0000] [78736] [INFO] Booting worker with pid: 78736
[2024-08-11 23:17:47 +0000] [78741] [INFO] Booting worker with pid: 78741

Aborted!
[2024-08-11 23:21:57 +0000] [78733] [INFO] Handling signal: int
[2024-08-11 23:21:57 +0000] [78734] [INFO] Worker exiting (pid: 78734)
[2024-08-11 23:21:57 +0000] [78735] [INFO] Worker exiting (pid: 78735)
[2024-08-11 23:21:57 +0000] [78736] [INFO] Worker exiting (pid: 78736)
[2024-08-11 23:21:57 +0000] [78741] [INFO] Worker exiting (pid: 78741)
[2024-08-11 23:21:58 +0000] [78733] [INFO] Shutting down: Master
